pyo3-ffi/src/lib.rs

```diff
@@ -0,0 +1,455 @@
+pub use self::pythonrun::*;
+pub use self::rangeobject::*;
+pub use self::setobject::*;
+pub use self::sliceobject::*;
+pub use self::structseq::*;
+pub use self::sysmodule::*;
+pub use self::traceback::*;
+pub use self::tupleobject::*;
+pub use self::typeslots::*;
+pub use self::unicodeobject::*;
+pub use self::warnings::*;
+pub use self::weakrefobject::*;
+
+mod abstract_;
+// skipped asdl.h
+// skipped ast.h
+mod bltinmodule;
+mod boolobject;
+mod bytearrayobject;
+mod bytesobject;
+// skipped cellobject.h
+mod ceval;
+// skipped classobject.h
+mod code;
+mod codecs;
+mod compile;
+mod complexobject;
+#[cfg(all(Py_3_8, not(Py_LIMITED_API)))]
+mod context; // It's actually 3.7.1, but no cfg for patches.
+#[cfg(not(Py_LIMITED_API))]
+pub(crate) mod datetime;
+mod descrobject;
+mod dictobject;
+// skipped dynamic_annotations.h
+mod enumobject;
+// skipped errcode.h
+mod eval;
+// skipped exports.h
+mod fileobject;
+mod fileutils;
+mod floatobject;
+// skipped empty frameobject.h
+#[cfg(not(Py_LIMITED_API))]
+pub(crate) mod funcobject;
+// skipped genericaliasobject.h
+#[cfg(not(Py_LIMITED_API))]
+mod genobject;
+mod import;
+// skipped interpreteridobject.h
+mod intrcheck;
+mod iterobject;
+mod listobject;
+// skipped longintrepr.h
+mod longobject;
+pub(crate) mod marshal;
+mod memoryobject;
+mod methodobject;
+mod modsupport;
+mod moduleobject;
+// skipped namespaceobject.h
+mod object;
+mod objimpl;
+// skipped odictobject.h
+// skipped opcode.h
+// skipped osdefs.h
+mod osmodule;
+// skipped parser_interface.h
+// skipped patchlevel.h
+// skipped picklebufobject.h
+// skipped pyctype.h
+// skipped py_curses.h
+#[cfg(not(Py_LIMITED_API))]
+mod pyarena;
+mod pycapsule;
+// skipped pydecimal.h
+// skipped pydtrace.h
+mod pyerrors;
+// skipped pyexpat.h
+// skipped pyfpe.h
+mod pyframe;
+mod pyhash;
+mod pylifecycle;
+// skipped pymacconfig.h
+// skipped pymacro.h
+// skipped pymath.h
+mod pymem;
+mod pyport;
+mod pystate;
+mod pythonrun;
+// skipped pystrhex.h
+// skipped pystrcmp.h
+mod pystrtod;
+// skipped pythread.h
+// skipped pytime.h
+mod rangeobject;
+mod setobject;
+mod sliceobject;
+mod structseq;
+mod sysmodule;
+mod traceback;
+// skipped tracemalloc.h
+mod tupleobject;
+mod typeslots;
+mod unicodeobject;
+mod warnings;
+mod weakrefobject;
+
+// Additional headers that are not exported by Python.h
+pub mod structmember;
+
+// "Limited API" definitions matching Python's `include/cpython` directory.
+#[cfg(not(Py_LIMITED_API))]
+mod cpython;
+
+#[cfg(not(Py_LIMITED_API))]
+pub use self::cpython::*;
+
+/// This cell type allows shared mutability for python types.
+/// It is intended to hold raw pointers to the python heap and therefore only accepts copy types to
+/// reduce the changes of accidental undefined behaviour.
+/// Thread safety is ensured by the python GIL.
+/// Therefore accessing the contents of this cell in any form while the GIL is not held is UB
+#[repr(transparent)]
+pub struct UnsafeGILCell<T>(Cell<*mut T>);
```

Comment:
Two thoughts:
- `Cell` just is a wrapper around `UnsafeCell`, so I'd prefer we use `UnsafeCell` as the inner type.
- I don't think that we need a generic type here, given it's only used once in the . Also, we don't really need to give downstream users the possibility to write to the datetime pointer. Let me suggest something directly in the `datetime` bindings...

Reply:
My Intention with the pub cell was to let downstream crates (of pyo3_ffi) use the same cell. While this is trivial to implement it would be a nice QOL addition because having mutable statics for exactly the same purpose as here (loading some python type/api capsule) is a pretty common pattern if one uses raw ffi. However I your solution below is much nicer and in the end my UnsafeGILCell is probably not gonna cover all needs of downstream crates anyway and those crates should probably just use similar wrapper types instead